feat(istio): Add Istio image build infrastructure

electricjesus opened the pull request:
## Description

- Adds build infrastructure for Istio components: ztunnel, pilot, install-cni, and proxyv2                                                                                                                                                         - Includes Semaphore CI/CD pipelines and Dockerfiles                                                                                                                                                                                               
- Cherry-picked from enterprise repo

## Related issues/PRs

<!-- If appropriate, include a link to the issue this fixes.
fixes <ISSUE LINK>

If appropriate, add links to any number of PRs documented by this PR
documents <PR LINK>
-->

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
added istio image build infrastructure
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

hjiawei committed:
Merge pull request #10181 from electricjesus/third_party-istio--ztunnel

feat(third_party): ztunnel component with ci/cd:

electricjesus committed:
feat(third_party): add istio-cni and istio-pilot build infrastructure

Cherry-picked from tigera/calico-private PR #10291

This adds:
- istio/ directory with Dockerfile-cni, Dockerfile-pilot, Dockerfile-ztunnel
- istio/Makefile for building Istio components
- SemaphoreCI pipeline for building and pushing istio images
- Patches for DSCP magic mark support

electricjesus committed:
feat(istio): Add tigera/istio-proxyv2 component

Cherry-picked from tigera/calico-private PR #10321

This adds:
- istio/Dockerfile-proxyv2 for building proxyv2 image
- istio/Dockerfile-install-cni for CNI installation
- Updated Makefile with proxyv2 targets
- Updated semaphore blocks for proxyv2 builds

electricjesus committed:
fix(istio): Remove istio-proxyv2 image in istio clean target

Cherry-picked from tigera/calico-private PR #10334

electricjesus committed:
chore(istio): upgrade Istio components to 1.28.1

Cherry-picked from tigera/calico-private PR #10343

This updates:
- Istio version to 1.28.1
- Patches for DSCP magic mark support
- Dependencies updates

electricjesus committed:
fix(istio-install-cni): add missing CNI dependencies

Cherry-picked from tigera/calico-private PR #10361

electricjesus committed:
fix(istio-install-cni): add TARGETPLATFORM arg

Cherry-picked from tigera/calico-private PR #10362

electricjesus committed:
fix(istio): enable cross-arch builds and use explicit file paths

Cherry-picked from tigera/calico-private PR #10365

electricjesus committed:
fix(istio): Use f1-standard-2 for istio cut release pipeline

Cherry-picked from tigera/calico-private PR #10376

electricjesus committed:
fix(istio): missing istio-cni itself

Cherry-picked from tigera/calico-private PR #10378

electricjesus committed:
fix(istio-proxyv2): add pilot agent

Cherry-picked from tigera/calico-private PR #10381

electricjesus committed:
fix(istio-proxyv2): use UBI10 base image to fix GLIBC compatibility

Cherry-picked from tigera/calico-private PR #10402

electricjesus committed:
fix(istio): create bin directory before copying LICENSE

The bin/LICENSE rule was failing in CI because the bin/ directory
did not exist when building istio-ztunnel (which has no binary build
dependency that would create the directory first).

electricjesus committed:
temp: temporarily delete other blocks to save resources

electricjesus committed:
fix(istio): update CI pipeline for open-source istio-ztunnel builds

- Add DOCKER_BUILD_THIRD_PARTY macro to lib.Makefile
- Add CALICO_BASE_UBI10 variable and build-arg support
- Update CALICO_BASE_UBI10_VER to use existing image tag
- Remove duplicate 50-istio-ztunnel.yml block
- Export BRANCH_NAME in semaphore blocks for image tagging
- Use docker.io/calico as DEV_REGISTRIES (rely on global docker-hub secret)
- Remove GCR login from Istio block (no longer needed)
- Update Dockerfile-ztunnel to pull from calico/istio-ztunnel:${BRANCH_NAME}
- Update Makefile to pass BRANCH_NAME build arg
- Skip istio-ztunnel image build on PRs (image not in registry yet)
- Remove private registry overrides from third_party/istio-ztunnel/Makefile

hjiawei commented on lib.Makefile:
```diff
@@ -943,6 +970,26 @@ retag-build-image-arch-with-registry-%: var-require-all-REGISTRY-BUILD_IMAGE-IMA
+retag-third-party-image-arch-with-registry-%: var-require-all-REGISTRY-THIRD_PARTY_IMAGE-IMAGETAG
+	docker pull $(THIRD_PARTY_REGISTRY)/$(THIRD_PARTY_IMAGE):$(LATEST_IMAGE_TAG)-$*
+	docker tag $(THIRD_PARTY_REGISTRY)/$(THIRD_PARTY_IMAGE):$(LATEST_IMAGE_TAG)-$* $(call filter-registry,$(REGISTRY))$(THIRD_PARTY_IMAGE):$(IMAGETAG)-$*
+
```
These Makefile targets are not needed.

hjiawei commented on istio/patches/0001-dscp-magic-mark-for-transparent-policies.patch:
Do we still need this patch? There appear to be two 0001 patch files with very similar content.

electricjesus committed:
refactor(istio): address review comments and cleanup infrastructure

- Remove unused istio/deps.txt file
- Remove cd-third-party-common and retag-third-party-* Makefile targets
- Change istio-ztunnel push registry from GCR/Quay to docker.io/calico
- Rename ZTUNNEL_IMAGE to istio-ztunnel-base to avoid overwriting final image
- Delete duplicate patch file (0001-dscp-magic-mark-for-transparent-policies.patch)
- Fix regex patterns in promotion config (release- -> release-.*)
- Regenerate semaphore.yml and semaphore-scheduled-builds.yml

electricjesus commented:
thanks @hjiawei ! I addressed your comments. I tried disabling the `if [ -z "${SEMAPHORE_GIT_PR_NUMBER}" ]; then make -C third_party/istio-ztunnel cd ARCHES=arm64 CONFIRM=true; fi` bit so that I can try if it does push to docker.io/calico, but it doesn't. perhaps the repo path needs to be created first?
